### PR TITLE
Rustup to 2016-03-11

### PR DESCRIPTION
**Pull request by mcarton**
Ref https://github.com/rust-lang/rust/pull/31710.
For tomorrow’s nightly, don’t bother building it today Travis.

### CODE REVIEW

**Comment by llogiq** [comment]
Travis did bother anyway :smiley:

**Comment by Manishearth** [comment]
New failures:

```
   Compiling semver v0.2.3

   Compiling clippy v0.0.49 (file:///home/travis/build/Manishearth/rust-clippy)

src/cyclomatic_complexity.rs:128:21: 128:32 error: unresolved enum variant, struct or const `TyFnDef` [E0419]

src/cyclomatic_complexity.rs:128                     ty::TyFnDef(_, _, ty) | ty::TyFnPtr(ty) if ty.sig.skip_binder().output.diverges() => {

                                                     ^~~~~~~~~~~

src/cyclomatic_complexity.rs:128:21: 128:32 help: run `rustc --explain E0419` to see a detailed explanation

src/cyclomatic_complexity.rs:128:45: 128:56 error: unresolved enum variant, struct or const `TyFnPtr` [E0419]

src/cyclomatic_complexity.rs:128                     ty::TyFnDef(_, _, ty) | ty::TyFnPtr(ty) if ty.sig.skip_binder().output.diverges() => {

                                                                             ^~~~~~~~~~~

src/cyclomatic_complexity.rs:128:45: 128:56 help: run `rustc --explain E0419` to see a detailed explanation

src/eta_reduction.rs:63:21: 63:32 error: unresolved enum variant, struct or const `TyFnDef` [E0419]

src/eta_reduction.rs:63                     ty::TyFnDef(_, _, fn_ty) | ty::TyFnPtr(fn_ty) => {

                                            ^~~~~~~~~~~

src/eta_reduction.rs:63:21: 63:32 help: run `rustc --explain E0419` to see a detailed explanation

src/eta_reduction.rs:63:48: 63:59 error: unresolved enum variant, struct or const `TyFnPtr` [E0419]

src/eta_reduction.rs:63                     ty::TyFnDef(_, _, fn_ty) | ty::TyFnPtr(fn_ty) => {

                                                                       ^~~~~~~~~~~

src/eta_reduction.rs:63:48: 63:59 help: run `rustc --explain E0419` to see a detailed explanation

src/derive.rs:158:29: 158:54 error: no associated item named `TyFnPtr` found for type `rustc::middle::ty::sty::TypeVariants<'_>` in the current scope

src/derive.rs:158                             TypeVariants::TyFnPtr(..) => {

                                              ^~~~~~~~~~~~~~~~~~~~~~~~~

src/mut_reference.rs:57:9: 57:49 error: no associated item named `TyFnDef` found for type `rustc::middle::ty::sty::TypeVariants<'_>` in the current scope

src/mut_reference.rs:57         TypeVariants::TyFnDef(_, _, ref fn_type) | TypeVariants::TyFnPtr(ref fn_type) => {

                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/mut_reference.rs:57:52: 57:86 error: no associated item named `TyFnPtr` found for type `rustc::middle::ty::sty::TypeVariants<'_>` in the current scope

src/mut_reference.rs:57         TypeVariants::TyFnDef(_, _, ref fn_type) | TypeVariants::TyFnPtr(ref fn_type) => {

                                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/mut_reference.rs:58:31: 58:42 error: the type of this value must be known in this context

src/mut_reference.rs:58             let parameters = &fn_type.sig.skip_binder().inputs;

                                                      ^~~~~~~~~~~

src/mut_reference.rs:59:63: 59:80 error: the type of this value must be known in this context

src/mut_reference.rs:59             for (argument, parameter) in arguments.iter().zip(parameters.iter()) {

                                                                                      ^~~~~~~~~~~~~~~~~

src/mut_reference.rs:60:23: 60:36 error: the type of this value must be known in this context

src/mut_reference.rs:60                 match parameter.sty {

                                              ^~~~~~~~~~~~~

src/mut_reference.rs:63:60: 63:73 error: the type of this value must be known in this context

src/mut_reference.rs:63                         if let ExprAddrOf(MutMutable, _) = argument.node {

                                           
```

**Comment by mcarton** [comment]
@Manishearth patience, the right nightly hasn’t been release yet, :stuck_out_tongue: 
Travis has
```rust
$ rustc --version
rustc 1.9.0-nightly (cc62db802 2016-03-09)
```
https://github.com/rust-lang/rust/pull/31710 was merged yesterday (2016-03-**10**) at 7:41 AM GTM+1.

File: src/mut_reference.rs [hunk 1]
```diff
@@ -53,21 +53,24 @@ impl LateLintPass for UnnecessaryMutPassed {
...
-                                  argument.span,
-                                  &format!("The function/method \"{}\" doesn't need a mutable reference", name));
+    match type_definition.sty {
+        TypeVariants::TyFnDef(_, _, ref fn_type) | TypeVariants::TyFnPtr(ref fn_type) => {
+            let parameters = &fn_type.sig.skip_binder().inputs;
+            for (argument, parameter) in arguments.iter().zip(parameters.iter()) {
+                match parameter.sty {
+                    TypeVariants::TyRef(_, TypeAndMut {mutbl: MutImmutable, ..}) |
+                    TypeVariants::TyRawPtr(TypeAndMut {mutbl: MutImmutable, ..}) => {
+                        if let ExprAddrOf(MutMutable, _) = argument.node {
```
Comment (eddyb) [inline src/mut_reference.rs]: Why is this code using `TypeVariants` directly?

File: src/mut_reference.rs [hunk 2]
```diff
@@ -53,21 +53,24 @@ impl LateLintPass for UnnecessaryMutPassed {
 }
 
 fn check_arguments(cx: &LateContext, arguments: &[P<Expr>], type_definition: &TyS, name: &str) {
```
Comment (eddyb) [inline src/mut_reference.rs]: `&TyS` is `Ty`...

File: src/mut_reference.rs [hunk 3]
```diff
@@ -53,21 +53,24 @@ impl LateLintPass for UnnecessaryMutPassed {
...
-                TypeVariants::TyRef(_, TypeAndMut {mutbl: MutImmutable, ..}) |
-                TypeVariants::TyRawPtr(TypeAndMut {mutbl: MutImmutable, ..}) => {
-                    if let ExprAddrOf(MutMutable, _) = argument.node {
-                        span_lint(cx,
-                                  UNNECESSARY_MUT_PASSED,
-                                  argument.span,
-                                  &format!("The function/method \"{}\" doesn't need a mutable reference", name));
+    match type_definition.sty {
+        TypeVariants::TyFnDef(_, _, ref fn_type) | TypeVariants::TyFnPtr(ref fn_type) => {
+            let parameters = &fn_type.sig.skip_binder().inputs;
```
Comment (eddyb) [inline src/mut_reference.rs]: You can use `.fn_sig()` to get to this point without matching.

Comment (mcarton) [inline src/mut_reference.rs]: This panics for non-fn types:
```
thread 'rustc' panicked at 'Ty::fn_sig() called on non-fn type: Box<core::ops::Fn(collections::vec::Vec<u32>) + 'static>', ../src/librustc/middle/ty/sty.rs:1144
```
with I guess:
```rust
pub fn test2(foo: Box<Fn(Vec<u32>)>) { // pass if #31 is fixed
    foo(vec![1, 2, 3])
}
```
The matching just ignores them.

File: src/mut_reference.rs (hunk 1 above)
Comment (mcarton) [inline src/mut_reference.rs]: No idea, I was like that. I don’t dislike it. I left it.

File: src/mut_reference.rs (hunk 3 above)
Comment (eddyb) [inline src/mut_reference.rs]: That's not a function call... that's an overloaded method and you need to do something completely different based on whether the method map has an entry. You should really handle that correctly, perhaps abstracting over it.

File: src/mut_reference.rs (hunk 1 above)
Comment (eddyb) [inline src/mut_reference.rs]: Right, but, it *should* be using `ty`.

**Pull request merged**